catalog/secrets/cache/src/main/java/org/projectnessie/catalog/secrets/cache/CachingSecretsBackend.java

```diff
@@ -0,0 +1,260 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.catalog.secrets.cache;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.stream.Collectors.toMap;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import com.google.common.annotations.VisibleForTesting;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.binder.cache.CaffeineStatsCounter;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.TimeUnit;
+import java.util.function.LongSupplier;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.projectnessie.catalog.secrets.spi.SecretsSupplier;
+
+public class CachingSecretsBackend {
+
+  private static final Map<String, String> CACHE_NEGATIVE_SENTINEL = new HashMap<>();
+
+  public static final String CACHE_NAME = "nessie-secrets";
+  private static final long NOT_CACHED = 0L;
+
+  @VisibleForTesting final Cache<CacheKeyValue, Map<String, String>> cache;
+  private final long ttlNanos;
+  private final LongSupplier clock;
+
+  public CachingSecretsBackend(SecretsCacheConfig config) {
+    OptionalLong ttl = config.ttlMillis();
+    this.ttlNanos = ttl.isPresent() ? TimeUnit.MILLISECONDS.toNanos(ttl.getAsLong()) : 0L;
+    this.clock = config.clockNanos();
+
+    Caffeine<CacheKeyValue, Map<String, String>> cacheBuilder =
+        Caffeine.newBuilder()
+            .expireAfter(
+                new Expiry<CacheKeyValue, Map<String, String>>() {
+                  @Override
+                  public long expireAfterCreate(
+                      CacheKeyValue key, Map<String, String> value, long currentTimeNanos) {
+                    long expire = key.expiresAtNanosEpoch;
+                    if (expire == NOT_CACHED) {
+                      return 0L;
+                    }
+                    long remaining = expire - currentTimeNanos;
+                    return Math.max(0L, remaining);
+                  }
+
+                  @Override
+                  public long expireAfterUpdate(
+                      CacheKeyValue key,
+                      Map<String, String> value,
+                      long currentTimeNanos,
+                      @NonNegative long currentDurationNanos) {
+                    return expireAfterCreate(key, value, currentTimeNanos);
```

Comment:
nit: why not simply return `currentDurationNanos`? AFAIK, it is the value returned by `expireAfterCreate` at entry creation time.